jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandler.java

```diff
@@ -2591,27 +2592,23 @@ public ClassLoader getClassLoader()
             {
                 // check to see if the classloader of the caller is the same as the context
                 // classloader, or a parent of it, as required by the javadoc specification.
-                try
-                {
-                    Class<?> caller = new Caller().getCallerClass(2);
-                    boolean ok = false;
-                    ClassLoader callerLoader = caller == null ? null : caller.getClassLoader();
-                    while (!ok && callerLoader != null)
-                    {
-                        if (callerLoader == _classLoader)
-                            ok = true;
-                        else
-                            callerLoader = callerLoader.getParent();
-                    }
 
-                    if (ok)
-                        return _classLoader;
-                }
-                catch (Exception e)
+                // Wrap so that only Jetty code requires the "createSecurityManager" permission.
+                Caller caller = AccessController.doPrivileged((PrivilegedAction<Caller>)Caller::new);
+                Class<?> callerClass = caller.getCallerClass(2);
+                boolean ok = false;
+                ClassLoader callerLoader = callerClass == null ? null : callerClass.getClassLoader();
+                while (!ok && callerLoader != null)
                 {
-                    LOG.warn("Unable to check classloader of caller",e);
+                    if (callerLoader == _classLoader)
+                        ok = true;
```

Comment:
Like this?

``` java
                // Wrap so that only Jetty code requires the "createSecurityManager" permission.
                Caller caller = AccessController.doPrivileged((PrivilegedAction<Caller>)Caller::new);
                ClassLoader callerLoader = caller.getCallerClassLoader(2);
                while (callerLoader != null)
                {
                    if (callerLoader == _classLoader)
                        return _classLoader;
                    callerLoader = callerLoader.getParent();
                }

                // No classloader match, check permission
                AccessController.checkPermission(new RuntimePermission("getClassLoader"));
                return _classLoader;
```

Reply:
Thumbs up!